From 3d9f66f9e8daad0e706de84e1b79396bc8023237 Mon Sep 17 00:00:00 2001
From: Michal Hocko <mhocko@suse.cz>
Date: Wed, 14 May 2014 10:56:59 +0200
Subject: [PATCH 1/2] memcg: deprecate memory.force_empty knob
Patch-mainline: not yet (expected 3.16)
References: bnc#878274

force_empty has been introduced primarily to drop memory before it gets
reparented on the group removal.  This alone doesn't sound fully justified
because reparented pages which are not in use can be reclaimed also later
when there is a memory pressure on the parent level.

Mark the knob CFTYPE_INSANE which tells the cgroup core that it shouldn't
create the knob with the experimental sane_behavior.  Other users will get
informed about the deprecation and asked to tell us more because I do not
expect most users will use sane_behavior cgroups mode very soon.

Anyway I expect that most users will be simply cgroup remove handlers
which do that since ever without having any good reason for it.

If somebody really cares because reparented pages, which would be dropped
otherwise, push out more important ones then we should fix the reparenting
code and put pages to the tail.

[akpm: s/pr_info/pr_info_once/]
[akpm: fix garbled printk text]
Signed-off-by: Michal Hocko <mhocko@suse.cz>
Acked-by: Johannes Weiner <hannes@cmpxchg.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>

---
 Documentation/cgroups/memory.txt |    3 +++
 mm/memcontrol.c                  |    4 ++++
 2 files changed, 7 insertions(+)

--- a/Documentation/cgroups/memory.txt
+++ b/Documentation/cgroups/memory.txt
@@ -378,6 +378,9 @@
   Because rmdir() moves all pages to parent, some out-of-use page caches can be
   moved to the parent. If you want to avoid that, force_empty will be useful.
 
+  Please note that this knob is considered deprecated and will be removed
+  in future.
+
 5.2 stat file
 
 memory.stat file includes following statistics
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -3933,6 +3933,10 @@
 
 int mem_cgroup_force_empty_write(struct cgroup *cont, unsigned int event)
 {
+	pr_info_once("%s (%d): memory.force_empty is deprecated and will be "
+		     "removed.  Let us know if it is needed in your usecase at "
+		     "linux-mm@kvack.org\n",
+		     current->comm, task_pid_nr(current));
 	return mem_cgroup_force_empty(mem_cgroup_from_cont(cont), true);
 }
 
